File: Algorithm_LevelOne/Permutations/Permutations.py
# 46. Permutations Python

class Solution(object):
    def permute(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        output = []
        def backtrack(first = 0):
            if first is n:
                output.append(nums[:])
            for i in range(first, n):
                nums[first], nums[i] = nums[i], nums[first]
                backtrack(first + 1)
                nums[first], nums[i] = nums[i], nums[first]
        n = len(nums)
        backtrack()
        return output

File: Algorithm_LevelOne/Permutations/Permutations.cc
// 46. Permutations C++
#include <vector>

class Solution {
public:
    std::vector<std::vector<int>> output;
    std::vector<std::vector<int>> permute(std::vector<int>& nums) {
        backtrack(0, nums, nums.size());
        return output;
    }
    void backtrack(int first, std::vector<int>& nums, int length) {
        if (first == length) output.emplace_back(nums);
        for (int i = first; i < length; i++) {
            int tmp = nums[first];
            nums[first] = nums[i];
            nums[i] = tmp;
            backtrack(first + 1, nums, length);
            nums[i] = nums[first];
            nums[first] = tmp;
        }
    }
};
